source/agora/common/BlockStorage.d

```diff
@@ -441,30 +457,79 @@ public class BlockStorage
 
     ***************************************************************************/
 
-    private void write (size_t pos, const ubyte[] data) @trusted
+    private void write (size_t pos, const ubyte[] data) nothrow @trusted
     {
-        assert(this.file !is null);
-        assert(this.getFileLength >= pos+data.length);
-
-        foreach (idx, ref e; data)
-            this.file[pos + idx] = e;
+        try
+        {
+            if (
+                (this.file !is null) &&
+                (pos / MapSize == this.file_index) &&
+                ((pos+data.length) / MapSize == this.file_index))
+            {
+                const size_t x0 = pos - this.file_base;
+                foreach (idx, e; data)
+                    this.file[x0+idx] = e;
+            }
+            else
+            {
+                foreach (idx, e; data)
+                    this.writeByte(pos + idx, e);
+            }
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("BlockStorage.write: ", ex);
+        }
     }
 
     /***************************************************************************
 
-        Get memory mapped file size
+        Read unsigned byte to the file.
+
+        Params:
+            pos = Position to read
 
         Returns:
-            Returns size of file if is mapped, or 0 if is not mapped
+            Returns unsigned bytes read
 
     ***************************************************************************/
 
-    private size_t getFileLength () @trusted
+    private ubyte readByte (size_t pos) nothrow @trusted
     {
-        if (this.file is null)
-            return 0;
+        ubyte data;
+        try
+        {
+            if (this.map(pos / MapSize))
+                data = this.file[pos - this.file_base];
+        }
+        catch (Exception ex)
+        {
+            writeLog("BlockStorage.readBytes: ", ex);
```

Comment:
So in case of a failure, it will return a ubyte with the value 0 and the error will be ignored?